### PR TITLE
✨ feat : search functionality implemented

### PR DESCRIPTION
**Pull request by arkojeet-bera**
> Search page skeleton UI created (Responsive). Fitter functionality integrated to work synchronously with search. 

> Every query and filters rely on the url (which we are treating as the `single source of knowledge`)

Requesting further testing and feedback for improvement! @Bismay5467

### CODE REVIEW

File: client/src/pages/Search/Search.tsx [hunk 1]
```diff
@@ -1,3 +1,32 @@
...
+import DrawerFilter from './Filter/DrawerFilter';
+import AsideFilter from './Filter/AsideFilter';
+import { useSearchParams } from 'react-router-dom';
+import { useEffect } from 'react';
+
 export default function Search() {
-  return <div>Search</div>;
+  const [searchParams, _] = useSearchParams({});
+  useEffect(() => {
+    console.log(searchParams);
```
Comment (Bismay5467) [inline client/src/pages/Search/Search.tsx]: No console logs

**Comment by Bismay5467** [comment]
@simply-arko I suppose this PR was meant to get a initial feedback on the feat so I am not going for a code review. The functionalities are working as expected. Keep pushing any related code to this PR and will do a code review at end. Till then I am converting this PR to a draft.

**Comment by ghost** [comment]
@Bismay5467 I basically want this page to be like this only, at the moment. As of now I'm not dwelling deep into design and look of this route (**That's why u may see some bad HTML being written**). The main objective is to make sure all the components are in sync with the `url search params`. And the container where all the search results are being displayed (as cards) should re-render every-time and fire an API call with respect to changes!

**Comment by Bismay5467** [comment]
@simply-arko yeah make sense! Bad html code is fine for now. But I would recommend to built the entire search functionality now along with all the filters and API calls as it would appear in the final version. Bad code can always be refactored later!! Even the UI can be worked upon in future.

File: client/src/utils/swr/fetcher.ts [hunk 2]
```diff
@@ -1,6 +1,8 @@
-const fetcher = async (url: string | Request | URL) => {
-  const response = await fetch(url);
-  return response.json();
-};
+import axios from 'axios';
 
+export const axiosInstance = axios.create({
+  baseURL: 'https://examarchive-1.onrender.com/api/v1/',
```
Comment (Bismay5467) [inline client/src/utils/swr/fetcher.ts]: Put this in .env file

File: client/src/pages/Search/Search.tsx [hunk 3]
```diff
@@ -2,30 +2,39 @@ import ResultCard from './ResultCard/ResultCard';
...
-          <p className="font-bold">Showing search results for: </p>
-          <p>Query: {searchParams.get('query')}</p>
-          <p>Filter1: {searchParams.get('filter1')}</p>
-          <p>Filter2: {searchParams.get('filter2')}</p>
-          <p>Filter3: {searchParams.get('filter3')}</p>
-        </div>
-        <ResultCard />
+      <div className="p-4 flex flex-col gap-y-4 sm:col-span-7">
+        {data?.data.map((res) => (
+          <Link to={`/preview/${res._id}`} target="_blank" key={res._id}>
```
Comment (Bismay5467) [inline client/src/pages/Search/Search.tsx]: - Declare the routes as a const in a separate constant file. If necessary, changing the names of one or more routes would be easy.
- I was expecting the route as '/filePreview' as declared in backend

File: client/src/pages/Search/Search.tsx [hunk 4]
```diff
@@ -2,30 +2,39 @@ import ResultCard from './ResultCard/ResultCard';
...
   const [searchParams, _] = useSearchParams({});
-  useEffect(() => {
-    console.log(searchParams);
-  }, [searchParams]);
+  const { data, isLoading, error } = useSearch(
+    searchParams.get('query') || '',
+    page
+  );
+  console.log(data);
+
```
Comment (Bismay5467) [inline client/src/pages/Search/Search.tsx]: As mentioned earlier, define the type of the hooks for improved type safety.

File: client/src/hooks/swr-hooks/useSearch.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import { SearchResponse } from '@/types/api-types';
+import useSWR from 'swr';
+
+export const useSearch = (searchPrams: string, page: number) => {
```
Comment (Bismay5467) [inline client/src/hooks/swr-hooks/useSearch.ts]: Use objects to accept the parameters of any function, eliminating the need to know the order of the parameters.

**Comment by ghost** [comment]
@Bismay5467 The search is functional with now. Please have a look

**Comment by ghost** [comment]
@Bismay5467 As promised I've `refactored` code. Please have a look!

File: client/src/components/Navbar/Navbar.tsx [hunk 6]
```diff
@@ -6,24 +6,23 @@ import { useForm, SubmitHandler } from 'react-hook-form';
...
+  const currentLocation = useLocation();
+
   const search = (query: string) => {
-    if (query.length === 0) {
-      setSearchParams({ query: '' });
-      return;
-    }
-    const url = createSearchParams({ query: query }).toString();
-    navigate(`/search?${url}`);
+    if (currentLocation.pathname === '/') navigate(`/search?`);
```
Comment (Bismay5467) [inline client/src/components/Navbar/Navbar.tsx]: Like we discussed define the routes in a const object during the UI design phase. Fr its fine.

File: client/src/types/index.ts [hunk 7]
```diff
@@ -1,3 +1,9 @@
+export interface FilterInputs {
+  ExamType?: string;
+  subjectName?: string;
+  year?: string;
+  sortFilter?: string;
+}
```
Comment (Bismay5467) [inline client/src/types/index.ts]: Just some suggestions :
- Prefix interfaces with I and types with T for better code readability.
- Define all the types for a specific feat in a single file. For ex : all types related to search may be defined in a file named 'search.ts'

PS: this review includes the code written in '/api-types/search.ts'

File: client/src/pages/Search/Search.tsx [hunk 8]
```diff
@@ -1,3 +1,92 @@
...
+  const getKey: SWRInfiniteKeyLoader = (pageIndex, previousPageData) => {
+    // TODO: Content of previousPageData needs further testing
+
+    if (previousPageData && !previousPageData.hasMore) return null;
+    return getRequestObj(pageIndex + 1);
+  };
+
+  const {
+    data: response,
+    setSize,
```
Comment (Bismay5467) [inline client/src/pages/Search/Search.tsx]: Okay this implementation looks fine. But IMO its bit complicated. We can simply define a state in a parent component and pass the state to its child components and to update the state we can pass the event handlers to the child component. The only communication that would happen is btw the state defined in parent component and the URL. IMO that implementation looks more clean.

Ref [this gist](https://gist.github.com/Bismay5467/854fedfb2ada3caf6ce5fe8d3fb6101a) for any confusion

PS: this review includes the code defined in `FilterForm` component.

**Comment by Bismay5467** [comment]
@simply-arko left some reviews. You may refactor the code now or later during the UI design phase. If that's the case, let me know, I will merge this PR for now.

**Comment by ghost** [comment]
@Bismay5467 I've refactored the code as per suggestions. Please review!

**Pull request merged**